File: app/src/main/java/com/krystianwsul/checkme/utils/time/TimePairPersist.java
package com.krystianwsul.checkme.utils.time;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.krystianwsul.checkme.utils.CustomTimeKey;

import junit.framework.Assert;

public class TimePairPersist implements Parcelable {
    @Nullable
    private CustomTimeKey mCustomTimeKey;

    @NonNull
    private HourMinute mHourMinute;

    public TimePairPersist(@NonNull HourMinute hourMinute) {
        mHourMinute = hourMinute;
    }

    private TimePairPersist(@Nullable CustomTimeKey customTimeKey, @NonNull HourMinute hourMinute) {
        mCustomTimeKey = customTimeKey;
        mHourMinute = hourMinute;
    }

    public TimePairPersist(@NonNull TimePair timePair) {
        mCustomTimeKey = timePair.mCustomTimeKey;

        if (timePair.mHourMinute != null)
            mHourMinute = timePair.mHourMinute;
        else
            mHourMinute = HourMinute.getNextHour().getSecond();
    }

    public void setCustomTimeKey(@NonNull CustomTimeKey customTimeKey) {
        mCustomTimeKey = customTimeKey;
    }

    public void setHourMinute(HourMinute hourMinute) {
        Assert.assertTrue(hourMinute != null);

        mCustomTimeKey = null;
        mHourMinute = hourMinute;
    }

    @NonNull
    public HourMinute getHourMinute() {
        return mHourMinute;
    }

    @Nullable
    public CustomTimeKey getCustomTimeKey() {
        return mCustomTimeKey;
    }

    @NonNull
    public TimePair getTimePair() {
        if (mCustomTimeKey != null)
            return new TimePair(mCustomTimeKey);
        else
            return new TimePair(mHourMinute);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeParcelable(mHourMinute, 0);
        if (mCustomTimeKey != null) {
            dest.writeInt(1);
            dest.writeParcelable(mCustomTimeKey, 0);
        } else {
            dest.writeInt(0);
        }
    }

    public static Parcelable.Creator<TimePairPersist> CREATOR = new Creator<TimePairPersist>() {
        @Override
        public TimePairPersist createFromParcel(Parcel source) {
            HourMinute hourMinute = source.readParcelable(HourMinute.class.getClassLoader());
            Assert.assertTrue(hourMinute != null);

            CustomTimeKey customTimeKey;
            if (source.readInt() == 1) {
                customTimeKey = source.readParcelable(CustomTimeKey.class.getClassLoader());
                Assert.assertTrue(customTimeKey != null);
            } else {
                customTimeKey = null;
            }

            return new TimePairPersist(customTimeKey, hourMinute);
        }

        @Override
        public TimePairPersist[] newArray(int size) {
            return new TimePairPersist[size];
        }
    };
}

File: app/src/main/java/com/krystianwsul/checkme/utils/time/TimePairPersist.kt
package com.krystianwsul.checkme.utils.time

import android.os.Parcel
import android.os.Parcelable
import com.krystianwsul.checkme.utils.CustomTimeKey
import kotlin.properties.Delegates

class TimePairPersist private constructor(var customTimeKey: CustomTimeKey?, _hourMinute: HourMinute) : Parcelable {

    companion object {

        @JvmField
        val CREATOR: Parcelable.Creator<TimePairPersist> = object : Parcelable.Creator<TimePairPersist> {
            override fun createFromParcel(source: Parcel): TimePairPersist {
                val hourMinute = source.readParcelable<HourMinute>(HourMinute::class.java.classLoader)!!

                val customTimeKey: CustomTimeKey? = if (source.readInt() == 1) {
                    source.readParcelable(CustomTimeKey::class.java.classLoader)!!
                } else {
                    null
                }

                return TimePairPersist(customTimeKey, hourMinute)
            }

            override fun newArray(size: Int) = arrayOfNulls<TimePairPersist>(size)
        }
    }

    var hourMinute by Delegates.observable(_hourMinute) { _, _, _ -> customTimeKey = null }

    val timePair get() = customTimeKey?.let { TimePair(it) } ?: TimePair(hourMinute)

    constructor(timePair: TimePair) : this(timePair.mCustomTimeKey, timePair.mHourMinute
            ?: HourMinute.getNextHour().second)

    override fun describeContents() = 0

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.run {
            writeParcelable(hourMinute, 0)
            if (customTimeKey != null) {
                writeInt(1)
                writeParcelable(customTimeKey, 0)
            } else {
                writeInt(0)
            }
        }
    }
}
